apisix/utils/batch-processor.lua

```diff
@@ -63,17 +63,36 @@ local function set_metrics(self, count)
 end
 
 
+local function slice_batch(batch, n)
+    local slice = {}
+    for i = n or 1, #batch, 1 do
+        slice[#slice+1] = batch[i]
+    end
+    return slice
+end
+
+
 function execute_func(premature, self, batch)
     if premature then
         return
     end
 
-    local ok, err = self.func(batch.entries, self.batch_max_size)
+    --- In case of "err" and a valid "n" batch processor considers, all n-1 entries have been
+    --- successfully consumed and hence reschedule the job for entries with index n to #entries
+    --- based on the current retry policy.
+    local ok, err, n = self.func(batch.entries, self.batch_max_size)
     if not ok then
-        core.log.error("Batch Processor[", self.name,
-                       "] failed to process entries: ", err)
+        if n then
+            core.log.error("Batch Processor[", self.name, "] failed to process entries [",
+                            #batch.entries + 1 - (n or -1), "/", #batch.entries ,"]: ", err)
```

Comment:
The `n` can't be `nil`, so `n or -1` is unnecessary.

Reply:
For it not to be nil, we have to update all existing instances of consumers. also, it's kind of enforcing the future implementation to return something like `false, err, 1` when the consumers consume the entries in one go and hit an error. Do we really need that sort of complication? I think it's better to leave that part to the consumers themselves like if they want to use the subtle feature or not. What do you say?

Reply:
Oh damn, got your point. It's the if block, haha